subprojects/wrapper/src/main/java/org/gradle/wrapper/Install.java

```diff
@@ -77,7 +82,48 @@ public File call() throws Exception {
         });
     }
 
-    private File getDistributionRoot(File distDir, String distributionDescription) {
+    private String calculateSha256Sum(File file)
+            throws Exception {
+        MessageDigest md = MessageDigest.getInstance("SHA-256");
+        InputStream fis = new FileInputStream(file);
+        int n = 0;
+        byte[] buffer = new byte[4096];
+        while (n != -1) {
+            n = fis.read(buffer);
+            if (n > 0) {
+                md.update(buffer, 0, n);
+            }
+        }
+        byte byteData[] = md.digest();
+
+        StringBuffer hexString = new StringBuffer();
+        for (int i=0; i < byteData.length; i++) {
+            String hex=Integer.toHexString(0xff & byteData[i]);
+            if (hex.length() == 1) {
+                hexString.append('0');
+            }
+            hexString.append(hex);
+        }
+
+        return hexString.toString();
+    }
+
+    private File getAndVerifyDistributionRoot(File localZipFile, String distributionSha256Sum,
+                                              File distDir, String distributionDescription)
+            throws Exception {
+        // check SHA-256 checksum verification
+        if (distributionSha256Sum != null) {
+            logger.log("Verifying " + localZipFile.getName());
+
+            if (!distributionSha256Sum.equals(calculateSha256Sum(localZipFile))) {
+                deleteDir(distDir);
```

Comment:
This makes me slightly uncomfortable. Might be better to just throw without deleting, and include this path in the error message so the user can delete if they want to.